src/java.net.http/share/classes/jdk/internal/net/http/Exchange.java

```diff
@@ -156,20 +157,25 @@ void connection(HttpConnection connection) {
                     this.closeRequested = false;
                 }
             }
-            if (closeRequested) closeConnection(connection);
+            if (closeRequested) closeConnection(connection, cause);
         }
 
-        void closeConnection() {
+        void closeConnection(Throwable error) {
             HttpConnection connection;
+            Throwable cause;
             synchronized (this) {
+                if ((cause = this.cause) == null) {
```

Comment:
let's extract the assignment to a separate statement. Code is a bit hard to read.